homeassistant/components/ihc/__init__.py

```diff
@@ -2,70 +2,110 @@
+    controller_id: str = str(entry.unique_id)
+    url: str = entry.data[CONF_URL]
+    username: str = entry.data[CONF_USERNAME]
+    password: str = entry.data[CONF_PASSWORD]
+    autosetup: bool = entry.data[CONF_AUTOSETUP]
+    ihc_controller: IHCController = IHCController(url, username, password)
+    if not await hass.async_add_executor_job(ihc_controller.authenticate):
         _LOGGER.error("Unable to authenticate on IHC controller")
         return False
-    controller_id: str = ihc_controller.client.get_system_info()["serial_number"]
-    # Store controller configuration
     hass.data.setdefault(DOMAIN, {})
     hass.data[DOMAIN][controller_id] = {
         IHC_CONTROLLER: ihc_controller,
-        CONF_INFO: controller_conf[CONF_INFO],
-        IHC_CONTROLLER_INDEX: controller_index,
     }
-    if controller_conf[CONF_AUTOSETUP] and not autosetup_ihc_products(
-        hass, config, ihc_controller, controller_id
-    ):
+    if not await setup_controller_device(hass, ihc_controller, entry):
         return False
-    get_manual_configuration(hass, config, controller_conf, controller_id)
-    # We only want to register the service functions once for the first controller
-    if controller_index == 0:
+    if autosetup:
+        await hass.async_add_executor_job(
+            autosetup_ihc_products, hass, ihc_controller, controller_id
+        )
+    await hass.async_add_executor_job(manual_setup, hass, controller_id)
+    for component in IHC_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+    entry.add_update_listener(async_update_options)
+    # We only wan to register service functions once, in case you have multiple controllers
+    if len(hass.data[DOMAIN]) == 1:
         setup_service_functions(hass)
     return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, IHC_PLATFORMS
+    )
+    if not unload_ok:
+        return False
+    controller_id = config_entry.unique_id
+    ihc_controller = hass.data[DOMAIN][controller_id][IHC_CONTROLLER]
+    ihc_controller.disconnect()
+    hass.data[DOMAIN].pop(controller_id)
+    if hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
+    return True
+
+
+async def async_update_options(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+    """Update options."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+async def setup_controller_device(
+    hass: HomeAssistant, ihc_controller: IHCController, entry: ConfigEntry
+) -> bool:
+    """Register the IHC controller as a Home Assistant device."""
+    # We must have a controller id, and cast the unique_id to a string.
+    # we know it is not None because it will always be set to the controller serial during setup
+    controller_id: str = str(entry.unique_id)
```

Comment:
According to the `dovalidate`method the serial is already a string

Reply:
The cast to str is there because entry.unique_id is str|None
mypy will complain if the cast is  not there. 
That is also why I have the comment above, because we know is it not None. 
I have changed the comment a bit to make it more clear.

Reply:
just do an assert entry.unique_id is not None to fix mypy complaining or use cast from the typing module

Reply:
Added the assert and remove the cast